webapp/src/main/java/com/box/l10n/mojito/service/tm/TMTextUnitStatisticService.java

```diff
@@ -0,0 +1,154 @@
+        List<Long> textUnitIds = new ArrayList<>();
+        textUnitStatistics.forEach(textUnitStatistic -> {
+            String statisticTextUnitMD5 = getTextUnitMd5(textUnitStatistic);
+            TextUnitDTO foundTextUnitDTO = textUnitDTOsForAssetAndLocaleByMD5.get(statisticTextUnitMD5);
+
+            if (foundTextUnitDTO != null) {
+                textUnitIds.add(foundTextUnitDTO.getTmTextUnitId());
+                md5ToTextUnitStatisticMap.put(statisticTextUnitMD5, textUnitStatistic);
+            } else {
+                logUnmatchedStatistics(asset, textUnitStatistic);
+            }
+        });
+
+        Map<ImportTextUnitStatisticsBody, TMTextUnit> statisticsToTextUnitsMap = getStatisticsToTextUnitMap(md5ToTextUnitStatisticMap, textUnitIds);
+        uploadStatistics(statisticsToTextUnitsMap);
+
+        return new PollableFutureTaskResult();
+    }
+
+    private void logUnmatchedStatistics(Asset asset, ImportTextUnitStatisticsBody textUnitStatistic) {
+        String message = String.format(
+                "No equivalent text unit found, skip import statistics for text unit with name: %1$s for asset with id: %2$s & path: %3$s",
+                textUnitStatistic.getName(),
+                asset.getId(),
+                asset.getPath());
+
+        logger.warn(message);
+
+        Tags tags = Tags.of("assetId", String.valueOf(asset.getId()),
+                "assetPath", asset.getPath());
+
+        meterRegistry.summary("services.TMTextUnitStatisticService.import_unmatched_statistics", tags);
+    }
+
+    private Map<ImportTextUnitStatisticsBody, TMTextUnit> getStatisticsToTextUnitMap(Map<String, ImportTextUnitStatisticsBody> md5ToTextUnitStatisticMap, List<Long> textUnitIds) {
+        List<TMTextUnit> textUnits = tmTextUnitRepository.findByIdInAndEagerFetchStatistics(textUnitIds);
+        return textUnits.stream().collect(
+                Collectors.toMap(tu -> md5ToTextUnitStatisticMap.get(tu.getMd5()), tu -> tu));
+    }
+
+    private void uploadStatistics(Map <ImportTextUnitStatisticsBody, TMTextUnit> statisticsToTextUnitsMap) {
+        List<TMTextUnitStatistic> updatedStatistics = new ArrayList<>();
+
+        statisticsToTextUnitsMap.forEach((textUnitStatistic, tmTextUnit) -> {
+            logger.debug("Add statistics for text unit with name: {}.", textUnitStatistic.getName());
+
+            TMTextUnitStatistic statistic = tmTextUnit.getStatistic();
+
+            if (statistic == null) {
+                statistic = new TMTextUnitStatistic();
+                statistic.setTMTextUnit(tmTextUnit);
+            }
+
+            statistic.setLastDayUsageCount(textUnitStatistic.getLastDayEstimatedVolume());
+            statistic.setLastPeriodUsageCount(textUnitStatistic.getLastPeriodEstimatedVolume());
+            statistic.setLastSeenDate(textUnitStatistic.getLastSeenDate());
+
+            updatedStatistics.add(statistic);
+        });
+
+        tmTextUnitStatisticRepository.saveAll(updatedStatistics);
```

Comment:
not worried about data size? Should we have batching?

Reply:
batching added